Add failedExpectations to the karma result object

astoilkov opened the pull request:
I am adding this proposal because I am currently building a karma reporter and I can't get access to the `failedExpectations` in order to make a better printing for the expects.

johnjbarton commented:
The failedExpectations are reformatted for better printing in the same file:
https://github.com/karma-runner/karma-jasmine/blob/master/src/adapter.js#L256
See the function `formatFailedStep`.

astoilkov commented:
Yes. I am aware of that. In order to better illustrate why I need the `failedExpectations` I will paste the code I wrote using the array so you see that collecting the data from the formatted text will be extremely hard.

```
result.failedExpectations.forEach(expectation => {
  let line = expectation.stack
    .split('\n')
    .find(line => line.trim().startsWith('at UserContext.'));
  let file = line.match(/\/base\/(.+)\)$/)[1];
  let filePath = file.match(/(^.+)\?/)[1];
  let lineNumber = parseFloat(file.match(/:([0-9]+):/)[1]);
  let fullPath = path.join(process.cwd(), 'test', 'karma', filePath);
  let fileContents = jetpack.read(fullPath);
  let lines = fileContents.split('\n');
  let testCode = extractTestCode(lines, lineNumber - 1);

  let text = '';
  text += terminalLink(
    chalk.white.bold(`#${totalFailed} ${result.suite.join(' | ')}`),
    'file://' + fullPath
  );
  text += '\n\n';
  text += chalk.white('Test:\n\n');
  text += testCode;
  text += '\n\n';
  text += chalk.white('Expected:\n\n');
  text += getValueString(expectation.expected);
  text += '\n\n';
  text += chalk.white('Actual:\n\n');
  text += getValueString(expectation.actual);
  results.push(text);
});
```

johnjbarton commented:
My concern is that `failedExpectations` is deliberately preprocessed in the browser to avoid breaking cases, eg dramatically large stack traces that bog down the tests. There isn't any limit to the content of `failedExpectations` AFAIK.

astoilkov commented:
@johnjbarton I am not sure I understand fully. What means avoid breaking cases? Is there something that can break if `failedExpectations` is too long? I thought that `failedExpectations` are missing from the object because you are trying to replicate the `karma-mocha` API?

johnjbarton commented:
I did not write this plugin, I'm just trying to ensure changes fit in to the design.  The design reformats the jasmine results rather than just passing it through. I think now that the reason was not for size cutoff.

I believe the design here is for the 'framework adapter' to send test results to the karma server for reporter plugins to consume. The test results could have extra information for specific frameworks because the reporters consume both generic info and extra info (they can be specific to a framework). But the reporters should not have to have different code for each browser, that's the job of the framework adapter that runs in the browser. For that reason the framework adapters reformat eg the call stacks which are not standard across browsers. 

I am sure you can achieve your goal by changes to the adapter with in the design rather than just sending the data through.
